### PR TITLE
feat: Multiple sensory modalities with the right probabilities in AnnubesEnv

### PR DESCRIPTION
**Pull request by cantordust**
- Added the ability to present multiple sensory modalities in parallel with specified occurrence probabilities.
- Fixed some edge cases in the constructor and the _new_trial method of AnnubesEnv.
- Added extra tests for AnnubesEnv to capture the new functionality.

### CODE REVIEW

File: neurogym/envs/annubes.py [hunk 1]
```diff
@@ -138,45 +183,89 @@ def _new_trial(self, **kwargs: Any) -> dict:  # type: ignore[override]
...
-            self.last_modality = None
+
+        stim_types_final = list(stim_types)
 
         self.trial = {
             "catch": catch,
-            "stim_type": stim_type,
-            "stim_value": stim_value,
+            "stim_types": stim_types_final,
+            "stim_values": [stim_values[k] for k in stim_types_final],
```
Comment (DaniBodor) [inline neurogym/envs/annubes.py:268]: does it need to be converted to a list?
if so, wouldn't this be simpler?

```suggestion
    self.trial = {
            "catch": catch,
            "stim_types": list(stim_types),
            "stim_values": [stim_values[k] for k in stim_types],
```

File: neurogym/utils/data.py [hunk 2]
```diff
@@ -55,12 +55,9 @@ def __init__(
...
             seq_len = 1000
 
-        obs_shape = env.observation_space.shape
-        action_shape = env.action_space.shape
-        if len(action_shape) == 0:
-            self._expand_action = True
-        else:
-            self._expand_action = False
+        obs_shape = tuple(env.observation_space.shape)  # type: ignore[arg-type]
+        action_shape = tuple(env.action_space.shape)  # type: ignore[arg-type]
```
Comment (DaniBodor) [inline neurogym/utils/data.py:59]: see [comment](https://github.com/neurogym/neurogym/pull/145/files#r1993239624) in PR #145.

This workaround is also different from the one in 145, so ultimately it would lead to conflicts. Better to just solve it properly once.

File: neurogym/utils/data.py [hunk 3]
```diff
@@ -55,12 +55,9 @@ def __init__(
...
+        self._expand_action = len(action_shape) == 0
```
Comment (DaniBodor) [inline neurogym/utils/data.py:60]: this formulation is shorter than the previous one, but I find it harder to read. What was the reason for changing it?

an alternative, which I think is short and more readable (and allows for `action_shape` to be None) would be:

```suggestion
        self._expand_action = False if action_shape else True
```

or

```suggestion
        self._expand_action = not action_shape
```

**Review by DaniBodor — CHANGES_REQUESTED** [review]
Thanks for this new feature. Seems like a good idea to allow multiple parallel inputs simultaneously.

I think that the class is becoming a bit overly complex at this point (I commented on that below somewhere as well). Do you think it would make sense to create a subclass of `AnnubesEnv` for this purpose instead?

Note:
I haven't looked at the new notebook or test file yet. Will open this review as is first and address those files later.

**Comment by DaniBodor** [comment]
Hi @cantordust . Is this PR ready for re-review? If so, please tag it as such, otherwise I am not sure if you are still in the process of making adjustments/responding to my comments.

I also learned from @gcroci2 that we can get auto reviews from Co-pilot nowadays. I asked for one, not sure if it's useful. Let's see :)

File: neurogym/envs/annubes.py [hunk 4]
```diff
@@ -11,19 +11,35 @@ class AnnubesEnv(TrialEnv):
...
             Defaults to [0.8, 0.9, 1].
         stim_time: Duration of each stimulus in ms.
             Defaults to 1000.
         catch_prob: Probability of catch trials in the session. Must be between 0 and 1 (inclusive).
             Defaults to 0.5.
+        exclusive: This effectively switches between two trial modes:
+                - True: Makes the modalities mutually exclusive, so they can be presented sequentially,
+                but not together.
+                - False: Depending on the probabilities provided for each modality, there could
+                    be an overlap where multiple modalities are presented at the same time.
```
[the remark this answers is not in the document]
Comment (cantordust) [reply neurogym/envs/annubes.py:44]: Changed:

```
A dictionary of stimulus types mapped to possible intensity values of each stimulus that is actually presented to the agent with non-zero probability. Defaults to None.
```

File: neurogym/envs/annubes.py [hunk 5]
```diff
@@ -62,11 +78,12 @@ class AnnubesEnv(TrialEnv):
 
     def __init__(
         self,
-        session: dict[str, float] | None = None,
-        stim_intensities: list[float] | None = None,
+        session: dict | None = None,
```
[the remark this answers is not in the document]
Comment (cantordust) [reply neurogym/envs/annubes.py:84]: I added them back `dict[str, int | float] | None`

File: neurogym/envs/annubes.py [hunk 6]
```diff
@@ -79,20 +96,40 @@ def __init__(
...
-            session = {"v": 0.5, "a": 0.5}
+            stim_intensities = {k: [0.8, 0.9, 1.0] for k in session}
+
+        # Create a dictionary for max_sequential if only an int is given
+        if isinstance(max_sequential, int):
+            max_sequential = {k: max_sequential for k in session}
+
         super().__init__(dt=dt)
-        self.session = {i: session[i] / sum(session.values()) for i in session}
+        self.session = {k: v for k, v in session.items() if v > 0.0}
```
[the remark this answers is not in the document]
Comment (cantordust) [reply neurogym/envs/annubes.py:128]: If the probability is `0`, the stimulus should not be present in the session at all - this is how the session is actually defined. Personally, I would assume that all the keys in the `session` attribute should correspond to modalities that are actually presented, and if the probability is `0` for any of them, they shouldn't be available at all. At least that's my rationale...

I also didn't want to assume how the users would come up with the probabilities, so I tried to cover all options. For instance, the probabilities might be computed on the fly by some function, which may well return a `0` (or multiple `0`s).

File: neurogym/envs/annubes.py [hunk 7]
```diff
@@ -138,45 +183,89 @@ def _new_trial(self, **kwargs: Any) -> dict:  # type: ignore[override]
...
-                available_modalities = [mod for mod in self.session if mod != self.last_modality]
-                stim_type = self._rng.choice(available_modalities)
+        if (
+            self.max_sequential is not None
+            and None in self.max_sequential
+            and self.sequential_count[None] == self.max_sequential[None]
+        ):
+            catch = False
+        else:
+            catch = len(available_modalities) == 0 or self._rng.random() < self.catch_prob
```
[the remark this answers is not in the document]
Comment (cantordust) [reply neurogym/envs/annubes.py:215]: `None` represents `catch` - I guess it's convenient to use `None` to ensure that it's different from all the other modalities.

File: neurogym/utils/data.py (hunk 2 above)
Comment (cantordust) [reply neurogym/utils/data.py:59]: Sync'd with #145

File: neurogym/utils/data.py (hunk 3 above)
Comment (cantordust) [reply neurogym/utils/data.py:60]: The reason for changing it is not for brevity but the fact that I don't like to rely on implicit conversion to `bool`. Is there a situation where `action_shape` would be `None`?

File: docs/examples/psychometric_curve.ipynb
[the remark this answers is not in the document]
Comment (cantordust) [reply docs/examples/psychometric_curve.ipynb:1]: Done!

**Comment by DaniBodor** [comment]
@cantordust , can you please resolve the merge conflicts before I proceed with reviewing

**Comment by gcroci2** [comment]
> @DaniBodor @gcroci2 Ready for a review.

Since @DaniBodor has already reviewed the majority of this, I’ve removed myself. That said, I noticed that the psychometric curve notebook seems to be invalid @cantordust. Feel free to tag me again if you think a second review is needed.

File: neurogym/envs/annubes.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (DaniBodor) [reply neurogym/envs/annubes.py]: I meant the description of the `exclusive` argument

File: neurogym/envs/annubes.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (DaniBodor) [reply neurogym/envs/annubes.py]: I understand. But why not just leave it with probability 0 in `self.session`. It doesn't influence execution to have it there right? 

I would leave it in because:
- If a user inputs 0 as a probability for whatever , I would rather assume they had a reason and leave it in (e.g. consistent plotting colors) than scrape it. 
- If a machine inputs 0, I'd rather leave it in, for easier troubleshooting/retracing what happened.

What is the downside of leaving 0-probability keys in session?

**Comment by DaniBodor** [comment]
@cantordust , as discussed, i've changed this back to draft. please activate the notebook again and ping me when it's ready for review

**Comment by cantordust** [comment]
@DaniBodor Superceded by #214 - we can retire this PR if you are OK with that.
